arctic/store/metadata_store.py

```diff
@@ -0,0 +1,229 @@
+from datetime import datetime as dt
+import logging
+
+import pandas as pd
+
+import bson
+import pymongo
+from ..decorators import mongo_retry
+from ..exceptions import NoDataFoundException
+
+from .bson_store import BSONStore
+
+logger = logging.getLogger(__name__)
+
+METADATA_STORE_TYPE = 'MetadataStore'
+
+class MetadataStore(BSONStore):
+    """
+    Metadata Store. This stores metadata with timestamps to allow temporal queries.
+
+    Entries are stored in the following format:
+        'symbol': symbol name
+        'metadata': metadata to be persisted
+        'start_time': when entry becomes effective
+        'end_time': (Optional) when entry expires. If not set, it is still in effect
+
+    For each symbol end_time of a entry should match start_time of the next one except for the current entry.
+    """
+
+    @classmethod
+    def initialize_library(cls, arctic_lib, hashed=True, **kwargs):
+        MetadataStore(arctic_lib)._ensure_index()
+        BSONStore.initialize_library(arctic_lib, hashed, **kwargs)
+
+    @mongo_retry
+    def _ensure_index(self):
+        self.create_index([('symbol', pymongo.ASCENDING), ('start_time', pymongo.DESCENDING)],
+                          unique=True, background=True)
+
+    def __init__(self, arctic_lib):
+        self._arctic_lib = arctic_lib
+        self._collection = self._arctic_lib.get_top_level_collection().metadata
+
+    @mongo_retry
+    def find_one(self, *args, **kwargs):
+        """
+        See http://api.mongodb.com/python/current/api/pymongo/collection.html#pymongo.collection.Collection.find_one
+        """
+        return self._collection.find_one(*args, **kwargs)
+
+    @mongo_retry
```

Comment:
Can you push find_one, find_one_and_update, find_one_and_delete down into BSON store? Quite useful generally.